src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt

```diff
@@ -0,0 +1,127 @@
+
+class TransportDeleteTransformsAction @Inject constructor(
+    transportService: TransportService,
+    val client: Client,
+    actionFilters: ActionFilters
+) : HandledTransportAction<DeleteTransformsRequest, BulkResponse>(
+    DeleteTransformsAction.NAME, transportService, actionFilters, ::DeleteTransformsRequest
+) {
+
+    override fun doExecute(task: Task, request: DeleteTransformsRequest, actionListener: ActionListener<BulkResponse>) {
+
+        // TODO: if metadata id exists delete the metadata doc else just delete transform
+
+        // Use Multi-Get Request
+        val getRequest = MultiGetRequest()
+        request.ids.forEach { id ->
+            getRequest.add(MultiGetRequest.Item(INDEX_MANAGEMENT_INDEX, id))
+        }
+
+        client.multiGet(getRequest, object : ActionListener<MultiGetResponse> {
+            override fun onResponse(response: MultiGetResponse) {
+                try {
+                    bulkDelete(response, request.ids, actionListener)
+                } catch (e: Exception) {
+                    actionListener.onFailure(e)
+                }
+            }
+
+            override fun onFailure(e: Exception) = actionListener.onFailure(e)
+        })
+    }
+
+    private fun bulkDelete(response: MultiGetResponse, ids: List<String>, actionListener: ActionListener<BulkResponse>) {
+        val enabledIDs = mutableListOf<String>()
+        val sourceMissingIDs = mutableListOf<String>()
+        val notTransform = mutableListOf<String>()
+
+        response.responses.forEach {
+            // TODO: Check if the source is actually transform document
+            if (it.response != null && it.response.isExists) {
+                if (it.response.isSourceEmpty) {
+                    sourceMissingIDs.add(it.id)
+                }
+                val source = it.response.source
+                if (!source.keys.contains("transform")) {
+                    notTransform.add(it.id)
+                }
+
+                val transform = source["transform"] as Map<String, Any>
+                val enabled = transform["enabled"] as Boolean
+                if (enabled) {
+                    enabledIDs.add(it.id)
+                }
+            }
+        }
+
+        if (notTransform.isNotEmpty()) {
+            actionListener.onFailure(ElasticsearchStatusException(
+                "[$notTransform] IDs are not transforms!", RestStatus.BAD_REQUEST
+            ))
+        }
+
+        if (sourceMissingIDs.isNotEmpty()) {
+            actionListener.onFailure(ElasticsearchStatusException(
+                "[$sourceMissingIDs] are missing their source documents!", RestStatus.NOT_FOUND
+            ))
+        }
+
+        if (enabledIDs.isNotEmpty()) {
+            actionListener.onFailure(ElasticsearchStatusException(
+                "[$enabledIDs] transform(s) are enabled, please disable them before deleting them", RestStatus.CONFLICT
+            ))
+        }
```

Comment:
I think this can be cleaned up a little - code, errorMessage, missing return statments

```
        val enabledIDs = mutableListOf<String>()
        val notTransform = mutableListOf<String>()

        response.responses.forEach {
            if (it.response.isExists) {
                val source = it.response.source
                val enabled = (source["transform"] as Map<*, *>?)?.get("enabled") as Boolean?
                if (enabled == null) {
                    notTransform.add(it.id)
                }
                if (enabled == true) {
                    enabledIDs.add(it.id)
                }
            }
        }

        if (notTransform.isNotEmpty()) {
            actionListener.onFailure(ElasticsearchStatusException(
                "$notTransform IDs are not transforms!", RestStatus.BAD_REQUEST
            ))
            return
        }

        if (enabledIDs.isNotEmpty()) {
            actionListener.onFailure(ElasticsearchStatusException(
                "$enabledIDs transform(s) are enabled, please disable them before deleting them", RestStatus.CONFLICT
            ))
            return
        }
```